pkgs/development/libraries/physics/geant4/default.nix

```diff
@@ -64,7 +69,10 @@ stdenv.mkDerivation rec {
 
   enableParallelBuilding = true;
   nativeBuildInputs =  [ cmake ];
-  buildInputs = [ clhep expat zlib libGLU libGL xlibsWrapper libXmu ]
+
+  # Because you compile against Geant4, it's useful to have all the libs you'd
+  # want available
+  propagatedBuildInputs = [ clhep expat zlib libGLU libGL xlibsWrapper libXmu findutils ]
```

Comment:
The `propagatedBuildInputs` has quite narrow scope of uses. First thing is that you need to understand that it only works in the nix builds (and nix-shell, which tries to emulate the build environment). So adding `findutils` to `propagatedBuildInputs` will work only in those situations, but not for a program that runs outside of a nix build. This is the reason why we prefer patching and wrapping [1].

Now regarding the throwing all libraries into `propagatedBuildInputs`, this is also not what we usually do. While it is true that some libraries may impose some transitive dependencies on their users, these cases are usually an exception. A typical example would be if a library A is written in C/C++ and contains an unconditional include of a header of a library B, then, for sure, B goes to the `propagatedBuildInputs` of A. Or other example could be if A's build system imposes a requirement to also statically link library B. We actually want to keep `propagatedBuildInputs` as small as possible to ensure that the build environments of the dependant builds are most minimal (performant) and most predictable (easy to track).

[1] see makeWrapper in https://nixos.org/nixpkgs/manual/#ssec-stdenv-functions

Reply:
I am not able to find the source of dependency on `findutils`. I tried removing it from the `nativeBuildInputs`:
```diff
diff --git a/pkgs/development/libraries/physics/geant4/default.nix b/pkgs/development/libraries/physics/geant4/default.nix
index 02f609be526..d96293516aa 100644
--- a/pkgs/development/libraries/physics/geant4/default.nix
+++ b/pkgs/development/libraries/physics/geant4/default.nix
@@ -72,7 +72,7 @@ stdenv.mkDerivation rec {
 
   # Because you compile against Geant4, it's useful to have all the libs you'd
   # want available
-  propagatedBuildInputs = [ clhep expat zlib libGLU libGL xlibsWrapper libXmu findutils ]
+  propagatedBuildInputs = [ clhep expat zlib libGLU libGL xlibsWrapper libXmu ]
     ++ stdenv.lib.optionals enableGDML [ xercesc ]
     ++ stdenv.lib.optionals enableXM [ motif ]
     ++ stdenv.lib.optionals enableQT [ qtbase ]
```
And I don't see any shell hook message:
```
% nix-shell -I nixpkgs=`pwd` -p geant4

[nix-shell:/tmp/pr78775]$ 
```

Reply:
I apologize.  That was quite sloppy on my part.  I did see the shell hook message, but only with overriding `enableQt = true`, and it seems to come from a different package.

Reply:
Ok, let's talk about `propagatedBuildInputs`

I did some digging.

When you use the Geant4Config.cmake file (to compile *against* Geant4) it searches for the following packages using find_dependency (depending on configuration), clhep, expat, zlib, xercesc, vecgeom, freetype, hdf5, x11, Qt5foo, opengl, motif, (xquartzgl apple only).

so I'll place just the ones matching those in propagatedBuildInputs, and the rest in buildInputs.

Reply:
I must say I'm surprised by what they do in their configuration file. Usually, those are supposed to hold the hardcoded values. Instead, they lookup bunch of dependencies, which, I guess, supposed to be used by the caller. This is strange.